### PR TITLE
Cocoon backend enhancement - from go to dart

### PR DESCRIPTION
**Pull request by keyonghan**
This PR focused on below three APIs:
/api/update-agent-health
/api/update-timeseries
/api/update-benchmark-targets (this one doesn't seem to be used from client side)

### CODE REVIEW

File: app_dart/lib/src/request_handlers/update_benchmark_targets.dart [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    final TimeSeries timeSeries =
+        await config.db.lookupValue<TimeSeries>(timeSeriesKey, orElse: () {
+      throw BadRequestException('No such task: ${timeSeriesKey.id}');
+    });
+    timeSeries.goal = goal;
+    timeSeries.baseline = baseline;
+
+    await config.db.withTransaction<void>((Transaction transaction) async {
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_benchmark_targets.dart]: Here too - you can use `datastore.db.commit(inserts: ...)` directly.

File: app_dart/lib/src/request_handlers/update_benchmark_targets.dart [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      transaction.queueMutations(inserts: <TimeSeries>[timeSeries]);
+      await transaction.commit();
+    });
+
+    return UpdateBenchmarkTargetsResponse(timeSeries);
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_benchmark_targets.dart]: And `Body.forJson()` again if you prefer.

File: app_dart/lib/src/request_handlers/update_timeseries.dart [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    extends ApiRequestHandler<UpdateTimeSeriesResponse> {
+  const UpdateTimeSeries(
+    Config config,
+    AuthenticationProvider authenticationProvider, {
+    @visibleForTesting
+        this.datastoreProvider = DatastoreService.defaultProvider,
+  }) : super(config: config, authenticationProvider: authenticationProvider);
+
+  final DatastoreServiceProvider datastoreProvider;
+  static const String timeSeriesKeyParam = 'TimeSeriesKey';
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_timeseries.dart]: Newline

File: app_dart/lib/src/request_handlers/update_timeseries.dart [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  static const String baselineParam = 'Baseline';
+  static const String taskNameParam = 'TaskName';
+  static const String labelParam = 'Label';
+  static const String unitParam = 'Unit';
+  static const String archivedParam = 'Archived';
+
+  @override
+  Future<UpdateTimeSeriesResponse> post() async {
+    checkRequiredParameters(
+        <String>[timeSeriesKeyParam, goalParam, baselineParam]);
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_timeseries.dart]: `taskNameParam`, `labelParam`, `unitParam`, too

File: app_dart/lib/src/request_handlers/update_timeseries.dart [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      throw BadRequestException('No such task: ${timeSeriesKey.id}');
+    });
+    timeSeries.goal = goal;
+    timeSeries.taskName = taskName;
+    timeSeries.label = label;
+    timeSeries.unit = unit;
+    timeSeries.baseline = baseline;
+    timeSeries.archived = archived;
+
+    await config.db.withTransaction<void>((Transaction transaction) async {
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_timeseries.dart]: Here too - you can use `datastore.db.commit(inserts: ...)` directly.

File: app_dart/lib/src/request_handlers/update_timeseries.dart [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+      transaction.queueMutations(inserts: <TimeSeries>[timeSeries]);
+      await transaction.commit();
+    });
+
+    return UpdateTimeSeriesResponse(timeSeries);
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_timeseries.dart]: `Body.forJson()`

**Comment by keyonghan** [comment]
Thanks for the detailed comments! How shall I kick off the tests?

File: app_dart/lib/src/request_handlers/update_agent_health.dart [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+  @override
+  Map<String, dynamic> toJson() {
+    return <String, dynamic>{
+      'Agent': agent.agentId,
+      'Healthy': agent.isHealthy,
+      'Details': agent.healthDetails,
+    };
+  }
+}
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_agent_health.dart]: Add newline at end of file

File: app_dart/lib/src/request_handlers/update_benchmark_targets.dart [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    try {
+      timeSeriesKey = keyHelper.decode(requestData[timeSeriesKeyParam]);
+    } catch (error) {
+      throw BadRequestException(
+          'Bad timeSeries key: ${requestData[timeSeriesKeyParam]}');
+    }
+
+    final TimeSeries timeSeries = await config.db.lookupValue<TimeSeries>(timeSeriesKey, orElse: () {
+        throw BadRequestException('Invalid time series Key: $timeSeriesKey');
+      });
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_benchmark_targets.dart]: Here too (comma after `}`)

File: app_dart/lib/src/request_handlers/update_benchmark_targets.dart [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  final TimeSeries timeSeries;
+
+  @override
+  Map<String, dynamic> toJson() {
+    return <String, dynamic>{
+      'Goal': timeSeries.goal,
+      'Baseline': timeSeries.baseline,
+    };
+  }
+}
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_benchmark_targets.dart]: Newline at end of file

File: app_dart/lib/src/request_handlers/update_timeseries.dart [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+      throw const BadRequestException('Missing required parameter: Label');
+    }
+    if (unit.isEmpty) {
+      throw const BadRequestException('Missing required parameter: Unit');
+    }
+
+    final TimeSeries timeSeries =
+        await datastore.db.lookupValue<TimeSeries>(timeSeriesKey, orElse: () {
+      throw BadRequestException('No such timeseries: ${timeSeriesKey.id}');
+    });
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_timeseries.dart]: Comma

File: app_dart/lib/src/request_handlers/update_timeseries.dart [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+  @override
+  Map<String, dynamic> toJson() {
+    return <String, dynamic>{
+      'Goal': timeSeries.goal,
+      'Baseline': timeSeries.baseline,
+      'Archived': timeSeries.archived,
+    };
+  }
+}
```
Comment (tvolkert) [inline app_dart/lib/src/request_handlers/update_timeseries.dart]: Newline

File: app_dart/test/request_handlers/update_timeseries_test.dart [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+      expect(response.timeSeries.archived, true);
+      expect(response.timeSeries.baseline, 4.0);
+      expect(response.timeSeries.goal, 0);
+      expect(response.timeSeries.label, 'flutter_repo_batch');
+      expect(response.timeSeries.taskName, 'analyzer_benchmark');
+      expect(response.timeSeries.unit, 's');
+    });
+  });
+}
```
Comment (tvolkert) [inline app_dart/test/request_handlers/update_timeseries_test.dart]: Newline

**Pull request merged**